File: backend/python/huge_file/3_pin_processor.py
# file: 3_pin_processor.py
# Process component PIN blocks, merge and clean descriptions

import sys # argv, exit
import re  # findall, DOTALL
import os  # remove, exists

def main():
    if len(sys.argv) < 2: sys.exit(1)
    in_p, out_p = sys.argv[1], sys.argv[2] if len(sys.argv) > 2 else None
    if out_p and os.path.exists(out_p): os.remove(out_p)
    with open(in_p, 'r', encoding='utf-8') as f: content = f.read()
    matches = re.findall(r'<<<([^_]+)_([^>]+?)>>> START\s+<<<NAME>>> = (.*?)\s+<<<\1_\2_DATASHEET>>> START\s+(.*?)\s+<<<\1_\2_DATASHEET>>> END\s+<<<\1_\2>>> END', content, re.DOTALL)
    data = {}
    for comp, pid, name, desc in matches:
        if desc.strip() and desc.strip() != "NO_DATA":
            data.setdefault((comp, pid), {'name': name.strip(), 'descs': []})['descs'].append(desc.strip())
    blocks = []
    for (comp, pid), d in data.items():
        unique = []
        for l in '\n\n'.join(d['descs']).split('\n'):
            if l not in unique and l.strip(): unique.append(l)
        blocks.append(f"<<<{comp}_{pid}>>> START\n[NAME] = {d['name']}\n[DATASHEET] START\n{'\n'.join(unique)}\n[DATASHEET] END\n<<<{comp}_{pid}>>> END")
    res = '\n\n'.join(blocks)
    if out_p:
        with open(out_p, 'w', encoding='utf-8') as f: f.write(res)
    else: print(res)

if __name__ == "__main__":
    main()

File: src/typescript/huge_file/3_pin_processor.ts
import fs from 'fs';

export function pinProcessor(inputPath: string, outputPath?: string): string {
  if (!fs.existsSync(inputPath)) return "";
  console.log(`[PIN_PROC] Processing pins from ${inputPath}`);
  const content = fs.readFileSync(inputPath, 'utf8');
  const regex = /<<<([^_]+)_([^>]+?)>>> START\s+<<<NAME>>> = (.*?)\s+<<<\1_\2_DATASHEET>>> START\s+([\s\S]*?)\s+<<<\1_\2_DATASHEET>>> END\s+<<<\1_\2>>> END/g;
  
  const records: Record<string, { name: string, descs: string[] }> = {};
  let m;
  let matchCount = 0;
  while ((m = regex.exec(content)) !== null) {
    const comp = m[1];
    const pid = m[2];
    const name = m[3].trim();
    const desc = m[4].trim();
    const key = `${comp}_${pid}`;

    if (desc && desc !== "NO_DATA") {
      if (!records[key]) records[key] = { name, descs: [] };
      records[key].descs.push(desc);
      matchCount++;
    }
  }

  console.log(`[PIN_PROC] Found ${matchCount} datasheet fragments for ${Object.keys(records).length} unique pins.`);

  const blocks: string[] = [];
  for (const [key, data] of Object.entries(records)) {
    const uniqueLines = Array.from(new Set(data.descs.join('\n\n').split('\n')))
      .filter(l => l.trim().length > 0);
    
    blocks.push(`<<<${key}>>> START\n[NAME] = ${data.name}\n[DATASHEET] START\n${uniqueLines.join('\n')}\n[DATASHEET] END\n<<<${key}>>> END`);
  }

  const result = blocks.join('\n\n');
  if (outputPath) {
    fs.writeFileSync(outputPath, result);
  }
  return result;
}
